pwiz_tools/Shared/PanoramaClient/PanoramaFolderBrowser.cs

```diff
@@ -0,0 +1,753 @@
+        public TreeNode Clicked { get; private set; }
+        public bool ShowSky { get; private set; }
+        public string Path { get; private set; }
+
+        public bool CurNodeIsTargetedMS { get; private set; }
+        public string State { get; private set; }
+        public PanoramaServer ActiveServer { get; private set; }
+        public bool Testing { get; private set; }
+        public int NodeCount { get; private set; }
+        public bool ShowWebDav { get; set; }
+        public string SelectedPath { get; set; }
+        public string SelectedUrl { get; set; }
+
+        /// <summary>
+        /// Builds the TreeView of folders and restores any
+        /// previous state of the TreeView
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void FolderBrowser_Load(object sender, EventArgs e)
+        {
+            if (ShowWebDav && SelectedPath != null)
+            {
+                ActiveServer = _serverList.FirstOrDefault();
+                InitializeTreeFromPath(treeView, SelectedPath, ActiveServer);
+                AddWebDavFolders(treeView.SelectedNode);
+                AddSelectedFiles(treeView.Nodes, EventArgs.Empty);
+
+            } else if (_serverList != null)
+            {
+                foreach (var server in _listServerFolders)
+                {
+                    InitializeFolder(treeView, _uploadPerms, true, server.Value, server.Key);
+                }
+
+                if (!string.IsNullOrEmpty(State))
+                {
+                    _restorer.RestoreExpansionAndSelection(State);
+                    _restorer.UpdateTopNode();
+                    AddSelectedFiles(treeView.Nodes, EventArgs.Empty);
+                }
+                else
+                {
+                    FolderPath = string.Empty;
+                    ActiveServer = _serverList.FirstOrDefault();
+                    treeView.TopNode.Expand();
+                }
+
+                if (SelectedPath != null)
+                {
+                    if (SelectedPath.EndsWith("/"))
+                    {
+                        SelectedPath = SelectedPath.Remove(SelectedPath.Length - 1);
+                    }
+                    var uriFolderTokens = SelectedPath.Split('/');
+                    SelectNode(uriFolderTokens.LastOrDefault());
+                }
+            }
+            else
+            {
+                InitializeTreeViewTest(_server, treeView, _folderJson);
+            }
+
+            if (treeView.SelectedNode != null)
+            {
+                if (ActiveServer != null)
+                {
+                    UpdateLinkLabel(treeView.SelectedNode);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Initializes the JSON that will be
+        /// used to build the TreeView of folders
+        /// </summary>
+        private void InitializeServers()
+        {
+            if (_serverList != null)
```

Comment:
When the entire body of a function is inside of a single "if" statement, I usually like to invert the if and put a "return". The only advantage of this is that the rest of the code in the function does not have to be indented as far:
```
if (_serverList == null) 
{
   return;
}
// rest of the function goes here

```